### ISSUE TITLE
too many gatherers when running in tests

### ISSUE THREAD

**Issue by srenatus** [lyft/protoc-gen-star#11]
Hey there.

Thanks for PGS, it's a very helpful library. 😃 

One issue we've run into is that it seems like [due to `RegisterPlugin` modifying a global variable](https://github.com/lyft/protoc-gen-star/blob/321a2dd04f1e689e8a50c0fda5b27e2d63a5fd75/generator.go#L80), when running multiple test cases through code generation (i.e., compiling a `CodeGeneratorRequest`, running a PGS-based module, and inspecting the `CodeGeneratorResponse`), we end up with too many gatherers. It's one per test case, so, the first case has the line
```
[gatherer] Initializing
```
once, the second test case's output has
```
[gatherer] Initializing
[gatherer] Initializing
```
and so one.
Each test case execution runs
```go
	pgs.Init(
		pgs.FileSystem(fs),
		pgs.ProtocInput(bytes.NewReader(inp)),
		pgs.ProtocOutput(outbuf),
		pgs.RequirePlugin("policy"),
		pgs.DebugMode(),
	).RegisterModule(policy.Policy()).
		RegisterPostProcessor(pgs.GoFmt()).
		Render()
```

This is leading (by as of now still a little mysterious ways) to flakey tests that fail every now and then, because the gatherer finds some message "not hydrated".

Am I doing this wrong? Is there something I can do to have the test cases' execution be more isolated? I'd love some pointers 😉 

Thanks again
Stephan

**Comment by rodaine** [comment]
Sorry for the late reply! Yeah unfortunately it's a limitation of protoc-gen-go since there's a global plugin registry and lookup table of entities. Long-term, we'll move away from relying on protoc-gen-go and this will be a non issue. For now, one alternative is to use modules instead, since they're scoped to the generator. Another, but more cumbersome choice, is to have the tests call into a compiled version of your plugin and run them as a black box. We do this for [protoc-gen-validate](https://github.com/lyft/protoc-gen-validate/tree/master/tests/harness) (though it uses modules, it's more of just to keep tests blackbox and consistent across the multiple supported languages).

**Comment by rodaine** [comment]
Fixed in #30
